common/src/main/java/org/conscrypt/ClientSessionContext.java

```diff
@@ -131,6 +137,29 @@ private NativeSslSession getSession(String host, int port) {
         return null;
     }
 
+    private void putSession(HostAndPort key, NativeSslSession session) {
+        synchronized (sessionsByHostAndPort) {
+            List<NativeSslSession> sessions = sessionsByHostAndPort.get(key);
+            if (sessions == null) {
+                sessions = new ArrayList<NativeSslSession>();
+                sessionsByHostAndPort.put(key, sessions);
+            }
+            sessions.add(session);
```

Comment:
Since servers may potentially hand you an unbounded number of sessions, it may be prudent to limit the number. Both for memory and because your logic below is O(N) in the number of sessions per host.

Also there's not much point in carrying around lots of TLS 1.2 sessions.

Reply:
The superclass (`AbstractSessionContext`) implements a maximum number of cached sessions (across all hosts), so we should be okay on that front.

Reply:
Two cases I could imagine wanting a per-host limit are:
* If you care about one host blowing all the other host's entries out of the cache
* Maybe there's some caller that knows they only connect to one or two hosts and, as a result, didn't bother to set a limit.

But just saying the global one covers it seems plausible enough if you don't care about those cases.